libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp

```diff
@@ -153,6 +143,82 @@ void AP_BattMonitor_UAVCAN::handle_battery_info_aux(const BattInfoAuxCb &cb)
     _has_battery_info_aux = true;
 }
 
+void AP_BattMonitor_UAVCAN::handle_mppt_stream(const MpptStreamCb &cb)
+{
+    const bool use_input_value = (uint32_t(_params._options.get()) & uint32_t(AP_BattMonitor_Params::Options::MPPT_Use_Input_Value)) != 0;
+    const float voltage = use_input_value ? cb.msg->input_voltage : cb.msg->output_voltage;
+    const float current = use_input_value ? cb.msg->input_current : cb.msg->output_current;
+
+    // use an invalid soc so we use the library calculated one
+    const uint8_t soc = 101;
+
+    // convert C to Kelvin
+    const float temperature_K = isnanf(cb.msg->temperature) ? 0 : cb.msg->temperature + C_TO_KELVIN;
+
+    update_interim_state(voltage, current, temperature_K, soc); 
+
+    if (!_is_mppt_packet_digital) {
+        _is_mppt_packet_digital = true;
+        set_bootup_powered_state();
+    }
+
+    mppt_check_and_report_faults(cb.msg->fault_flags);
+}
+
+void AP_BattMonitor_UAVCAN::mppt_check_and_report_faults(const uint8_t flags)
+{
+    if (_mppt_fault_flags == flags) {
+        // only report flags when they change
+        return;
+    }
+    _mppt_fault_flags = flags;
+
+    for (uint8_t fault_bit=0x01; fault_bit <= 0x08; fault_bit <<= 1) {
+        // this loop is to generate multiple messages if there are multiple concurrent faults, but also run once if there are no faults
+        if ((fault_bit & flags) != 0 || flags == 0) {
+            const MPPT_FaultFlags err = (flags == 0) ? MPPT_FaultFlags::NONE : (MPPT_FaultFlags)fault_bit;
+            GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Battery %u: Fault: %s", (unsigned)_instance+1, mppt_fault_string(err));
+        }
+        if (flags == 0) {
+            return;
+        }
+    }
+}
+
+void AP_BattMonitor_UAVCAN::update_interim_state(const float voltage, const float current, const float temperature, const uint8_t soc)
+{
+    WITH_SEMAPHORE(_sem_battmon);
+
+    _interim_state.voltage = voltage;
+    _interim_state.current_amps = current;
+    _soc = soc;
+
+    if (!isnanf(temperature) && temperature > 0) {
+        // Temperature reported from battery in kelvin and stored internally in Celsius.
+        _interim_state.temperature = temperature - C_TO_KELVIN;
+        _interim_state.temperature_time = AP_HAL::millis();
+    }
+
+    const uint32_t tnow = AP_HAL::micros();
+
+    if (!_has_battery_info_aux || _is_mppt_packet_digital) {
+        uint32_t dt = tnow - _interim_state.last_time_micros;
+
+        // update total current drawn since startup
+        if (_interim_state.last_time_micros != 0 && dt < 2000000) {
+            // .0002778 is 1/3600 (conversion to hours)
+            float mah = (float) ((double) _interim_state.current_amps * (double) dt * (double) 0.0000002778f);
```

Comment:
Just put (1.0 / 36000.0) instead of the magic constant. Or break it out into a const. I realize this is a copy paste, but if we are doing it, then let's try and improve it.

I'm also not sold on the use of doubles before casting back down here, it will be much slower on the F4 boards, and I'm not sure the accuracy wins here either, and if it's really that great we should be doing it for the analog monitor.

Reply:
No objection here. It's a copy/paste from here.

Reply:
this has totally been removed.

Reply:
We've left this as-is because it is existing code but I've raised an issue here to resolve it in the future https://github.com/ArduPilot/ardupilot/issues/19452